features/graph/provider/bsm/src/main/java/org/opennms/netmgt/graph/provider/bsm/AbstractBusinessServiceVertex.java

```diff
@@ -45,56 +48,82 @@
     }
 
     private final static String PROPERTY_LEVEL = "level";
-    private final static String PROPERTY_TYPE = "type";
+    protected final static String PROPERTY_TYPE = "type";
     private final static String PROPERTY_REDUCTION_KEYS = "reductionKeys";
     private final static String PROPERTY_IS_LEAF = "isLeaf";
-
-    /**
-     * Creates a new {@link AbstractBusinessServiceVertex}.
-     *  @param id the unique id of this vertex. Must be unique overall the namespace.
-     * @param label a human readable label
-     * @param level the level of the vertex in the Business Service Hierarchy. The root element is level 0.
-     */
-    protected AbstractBusinessServiceVertex(String id, String label, int level, Type type, boolean isLeaf,
-                                            Set<String> reductionKeys) {
-        super(NAMESPACE, id);
-        setLabel(label);
-        setLevel(level);
-        setType(type);
-        setIsLeaf(isLeaf);
-        setReductionKeys(reductionKeys);
+    
+    public AbstractBusinessServiceVertex(GenericVertex genericVertex) {
+        super(genericVertex);
+        // specific checks for this class:
+        Objects.requireNonNull(getLevel(), "getLevel() cannot be null.");
+        Objects.requireNonNull(isLeaf(), "isLeaf() cannot be null.");
+        Objects.requireNonNull(getType(), "getType() cannot be null.");
+        Objects.requireNonNull(getReductionKeys(), "getReductionKeys() cannot be null.");
     }
 
     public int getLevel() {
         return this.delegate.getProperty(PROPERTY_LEVEL);
     }
 
-    public void setLevel(int level){
-        this.delegate.setProperty(PROPERTY_LEVEL, level);
-    }
-
     public boolean isLeaf(){
         return this.delegate.getProperty(PROPERTY_IS_LEAF);
     }
 
-    public void setIsLeaf(boolean isLeaf){
-        this.delegate.setProperty(PROPERTY_IS_LEAF, isLeaf);
-    }
-
     public Type getType() {
         return this.delegate.getProperty(PROPERTY_TYPE);
     }
 
-    public void setType(Type type) {
-        this.delegate.setProperty(PROPERTY_TYPE, type);
-    }
-
     public Set<String> getReductionKeys() {
         return this.delegate.getProperty(PROPERTY_REDUCTION_KEYS);
     }
 
-    public void setReductionKeys(Set<String> reductionKeys) {
-        this.delegate.setProperty(PROPERTY_REDUCTION_KEYS, reductionKeys);// TODO MVR collections cannot be persisted
+    public final static AbstractBusinessServiceVertex from(GenericVertex genericVertex) {
```

Comment:
From what I can see, you could probably get rid of the concrete class GenericVertex and replace the few instances where you instantiate GenericVertex directly with a factory pattern for building the correct concrete type.

This `from` method should ideally just be taking an interface type as its parameter.

Once that is done you won't have an intermediate generic type and you shouldn't have to switch on its type like below. If the intent here is to still make a copy you could either add a copy method to each type or you could also do it via visitor pattern.

If there is any other cases similar to this in the PR we could solve them the same way. Otherwise the types just get too confusing in my opinion. Let me know if you want to discuss how to do this.

Reply:
The idea here is that GenericVertex and it's siblings GenericEdge, GenericGraph are the classes that hold the data and the ones we normally operate on. For convenience we can generate a domain specific view on those generic objects (such as ApplicationVertex) in order to perform domain specific operations. In this case (BSMGraph) however I am not even sure if we need all these domain classes.

Reply:
In that case couldn't `GenericVertex` be an interface that the domain specific vertices implement and get rid of the intermediate concrete class. We might also be able to get rid of `AbstractBusinessServiceVertex` as a concrete class and consolidate that into the interface.

When we instantiate a GenericVertex, at that time do we have the necessary information to create it as the appropriate domain vertex? I'm assuming yes since it seems that object and its properties are meant to be immutable. If we could do that we wouldn't have to derive the type from the properties it would already be typed correctly and we could visit it to get its type.

I'm just trying to see if we could avoid converting from one concrete type to another and work with interface types rather than a concrete type. As someone not that familiar with this code it looks like the object model is unnecessarily complicated, but I could be wrong.

Reply:
As mentioned earlier, this is a dummy implementation of the actual `BusinessServiceGraphProvider`. It was just copied over and its purpose is to just show data. However the proper implementation of this provider is still outstanding.